### PR TITLE
Wait for database capability to finish applying

### PR DESCRIPTION
**Pull request by schirevko**

### CODE REVIEW

File: mcs/db_cluster.go [hunk 1]
```diff
@@ -65,3 +65,31 @@ func databaseClusterStateRefreshFunc(client databaseClient, clusterID string) re
...
+			return nil, "", err
+		}
+		for _, i := range cluster.Instances {
+			capabilities, err := instanceGetCapabilities(client, i.ID).extract()
+			if err != nil {
+				return nil, "", err
+			}
+
+			for _, c := range capabilities {
+				if c.Status == string(dbCapabilityStatusApplying) {
```
Comment (ftersin) [inline mcs/db_cluster.go:83]: the case is impossible

File: mcs/db_instance.go [hunk 2]
```diff
@@ -139,6 +139,26 @@ func databaseInstanceStateRefreshFunc(client databaseClient, instanceID string)
...
 
+func databaseInstanceCapabilityStateRefreshFunc(client databaseClient, instanceID string) resource.StateRefreshFunc {
+	return func() (interface{}, string, error) {
+		capabilities, err := instanceGetCapabilities(client, instanceID).extract()
+		if err != nil {
+			return nil, "", err
+		}
+
+		for _, c := range capabilities {
+			if c.Status == string(dbCapabilityStatusApplying) {
```
Comment (ftersin) [inline mcs/db_instance.go:150]: ditto

File: mcs/resource_mcs_db_cluster.go [hunk 3]
```diff
@@ -405,6 +406,36 @@ func resourceDatabaseClusterCreate(d *schema.ResourceData, meta interface{}) err
...
+		applyCapabilityConf := &resource.StateChangeConf{
+			Pending:    []string{string(dbCapabilityStatusApplying)},
+			Target:     []string{string(dbCapabilityStatusActive)},
+			Refresh:    databaseClusterCapabilityStateRefreshFunc(DatabaseV1Client, cluster.ID),
+			Timeout:    d.Timeout(schema.TimeoutCreate),
+			Delay:      dbInstanceDelay,
+			MinTimeout: dbInstanceMinTimeout,
+		}
+		log.Printf("[DEBUG] Waiting for capabilities to become ready")
+		_, err = applyCapabilityConf.WaitForState()
```
Comment (ftersin) [inline mcs/resource_mcs_db_cluster.go:435]: nothing to wait here

File: mcs/resource_mcs_db_cluster.go [hunk 4]
```diff
@@ -405,6 +406,38 @@ func resourceDatabaseClusterCreate(d *schema.ResourceData, meta interface{}) err
 		if err != nil {
 			return fmt.Errorf("error applying capability to mcs_db_cluster %s: %s", cluster.ID, err)
 		}
+
+		// Wait until cluster changes status from CAPABILITY_APPLYING to ACTIVE
+		applyCapabilityClusterConf := &resource.StateChangeConf{
```
Comment (ftersin) [inline mcs/resource_mcs_db_cluster.go:411]: Could we move this waiting into a separate function?

File: mcs/resource_mcs_db_instance.go [hunk 5]
```diff
@@ -495,6 +490,48 @@ func resourceDatabaseInstanceCreate(d *schema.ResourceData, meta interface{}) er
...
+	if capabilities, ok := d.GetOk("capabilities"); ok {
+		capabilitiesOpts, err := extractDatabaseCapabilities(capabilities.([]interface{}))
+		if err != nil {
+			return fmt.Errorf("%s capability", message)
+		}
+		createOpts.Capabilities = capabilitiesOpts
+
+		var applyCapabilityOpts instanceApplyCapabilityOpts
+		applyCapabilityOpts.ApplyCapability.Capabilities = capabilitiesOpts
+		err = instanceAction(DatabaseV1Client, instance.ID, &applyCapabilityOpts).ExtractErr()
```
Comment (ftersin) [inline mcs/resource_mcs_db_instance.go:502]: Why we apply capabilities by separate request?

**Review by ftersin — APPROVED** [review]

**Review by irlndts — APPROVED** [review]

File: mcs/db_instance.go [hunk 6]
```diff
@@ -135,10 +135,52 @@ func databaseInstanceStateRefreshFunc(client databaseClient, instanceID string)
 			return i, i.Status, fmt.Errorf("there was an error creating the database instance")
 		}
 
+		if i.Status == "ACTIVE" {
```
Comment (pashapdev) [inline mcs/db_instance.go:138]: maybe for more consistency, use the **dbInstanceStatusActive** constant

File: mcs/db_cluster.go [hunk 7]
```diff
@@ -26,7 +26,7 @@ func flattenDatabaseClusterShard(inst dbClusterInstanceResp) map[string]interfac
 func getClusterStatus(c *dbClusterResp) string {
 	instancesStatus := string(dbInstanceStatusActive)
 	for _, inst := range c.Instances {
-		if inst.Status == "error" {
+		if inst.Status == "ERROR" {
```
Comment (pashapdev) [inline mcs/db_cluster.go:29]: maybe for more consistency, add the **dbInstanceStatusError** constant and use it

**Review by pashapdev — APPROVED** [review]

File: mcs/resource_mcs_db_instance.go [hunk 8]
```diff
@@ -27,14 +27,19 @@ const (
 )
 
 type dbInstanceStatus string
+type dbCapabilityStatus string
 
 var (
```
Comment (irlndts) [inline mcs/resource_mcs_db_instance.go:32]: i guess it's const

File: mcs/resource_mcs_db_instance.go [hunk 9]
```diff
@@ -27,14 +27,19 @@ const (
...
-	dbInstanceStatusDetach   dbInstanceStatus = "DETACH"
+	dbInstanceStatusDeleted            dbInstanceStatus   = "DELETED"
+	dbInstanceStatusBuild              dbInstanceStatus   = "BUILD"
+	dbInstanceStatusActive             dbInstanceStatus   = "ACTIVE"
+	dbInstanceStatusError              dbInstanceStatus   = "ERROR"
+	dbInstanceStatusShutdown           dbInstanceStatus   = "SHUTDOWN"
+	dbInstanceStatusResize             dbInstanceStatus   = "RESIZE"
+	dbInstanceStatusDetach             dbInstanceStatus   = "DETACH"
+	dbInstanceStatusCapabilityApplying dbInstanceStatus   = "CAPABILITY_APPLYING"
+	dbCapabilityStatusActive           dbCapabilityStatus = "ACTIVE"
```
Comment (irlndts) [inline mcs/resource_mcs_db_instance.go:41]: please split dbInstaceStatus and dbCapabilityStatus to independent blocks

File: mcs/resource_mcs_db_instance.go (hunk 8 above)
Comment (schirevko) [reply mcs/resource_mcs_db_instance.go:44]: I think we can write dbCapabilityStatus in the same vein as dbInstanceStatus and clusterStatus

**Pull request merged**